core/macros/src/main/scala/io/udash/macros/PropertyMacros.scala

```diff
@@ -101,40 +122,68 @@ class PropertyMacros(val c: blackbox.Context) extends MacroCommons {
-  private def doesReturnTypeForModelRequirements(symbol: Symbol, signatureType: Type): Boolean =
-    symbol.isMethod && isModelValue(symbol.typeSignatureIn(signatureType).resultType)
-
-  //Checks, if method is abstract, without parameters and returns ImmutableValue type
-  private def doesMethodMeetModelRequirements(s: Symbol, signatureType: Type): Boolean =
-    isAbstractMethod(s) && !takesParameters(s) && doesReturnTypeForModelRequirements(s, signatureType)
-
   //Returns filtered members without synthetic, top level and constructor of provided type
   private def filterMembers(valueType: Type) =
     valueType.members
       .filterNot(member => member.isSynthetic || isFromTopLevelType(member) || member.isConstructor || member.isType)
 
-  //Returns filtered memebrs which can be treated as Properties
-  private def propertyMembers(tpe: Type) =
+  //Checks, if method or val is abstract, without parameters and returns ImmutableValue type
+  private def doesMeetTraitModelRequirements(s: Symbol, signatureType: Type): Boolean =
+    doesMeetTraitElementsRequirements(s, signatureType) && isModelValue(s.typeSignatureIn(signatureType).resultType)
+
+  //Checks, if method or val is abstract, without parameters and returns ImmutableValue type
+  private def doesMeetTraitElementsRequirements(s: Symbol, signatureType: Type): Boolean =
+    (isAbstractMethod(s) && !takesParameters(s)) || isAbstractVal(s) ||
+      (s.isAbstract && s.isMethod && s.asMethod.isAccessor && s.asMethod.accessed == NoSymbol) // val in trait in scala 2.11
+
+  //Returns filtered memebrs which can be treated as Properties in trait based ModelProperty
+  private def traitBasedPropertyMembers(tpe: Type): Seq[Symbol] =
+    filterMembers(tpe).filter(!_.isPrivate)
+      .filter(s => doesMeetTraitModelRequirements(s, tpe)).toSeq
+
+  //Returns filtered memebrs which can be treated as Properties in trait based ModelProperty (with checking subtypes)
+  private def traitBasedPropertyPossibleMemebers(tpe: Type): Seq[Symbol] =
+    filterMembers(tpe).filter(!_.isPrivate)
+      .filter(s => doesMeetTraitElementsRequirements(s, tpe)).toSeq
+
+  //Returns filtered memebrs which can be treated as Properties in case class based ModelProperty
+  private def ccBasedPropertyMembers(tpe: Type): Seq[Symbol] =
     filterMembers(tpe)
-      .filter(s => doesMethodMeetModelRequirements(s, tpe))
+      .filter(m => !m.isPrivate && !(tpe <:< typeOf[Tuple2[_, _]] && m.name.decodedName.toString == "swap"))
+      .filter(m => m.isMethod && m.asMethod.isCaseAccessor).toSeq
 
   private def checkModel(tree: c.Tree): Boolean =
     c.typecheck(tree, silent = true) != EmptyTree
 
-  private def isModelPart(tpe: Type): Boolean = checkModel(q"""implicitly[$ModelPartCls[$tpe]]""")
+  private def checkModelWithCache(tree: c.Tree, tpe: Type, cache: mutable.Set[Any]): Boolean =
+    if (cache.contains(tpe)) true
+    else {
+      cache += tpe
+      val r = c.typecheck(tree, silent = true) != EmptyTree
```

Comment:
Can additional cache lookups be made during `c.typecheck`? If not, wouldn't it be better to just do `if (r) cache += tpe` later on?